Comorphisms/CASL2Prenex.hs

```diff
@@ -0,0 +1,184 @@
+                              foldl (\usedD vd@(Var_decl names sort r) -> 
+                                       if sort == s then -- found it, replace oN with nN in names
+                                         let vd' = Var_decl (map (\x -> if x == oN then nN else x) names) sort r  
+                                         in vd':usedD
+                                       else vd:usedD 
+                                 ) [] vds
+   checkAndReplace (n0, knownV, quantF, quants) (x, s) =
+      if (x,s) `elem` knownV 
+       then
+         let  
+           x' = genNumVar "x" n0
+           quants' = replVarDecl quants x x' s
+           quantF' = substitute x s (Qual_var x' s nullRange) quantF
+         in (n0 + 1, (x', s):knownV, quantF', quants')
+       else (n0    , (x , s):knownV, quantF , quants )  
+   (n'', vlist', form'', vdecls') = foldl checkAndReplace (n', vlist, form', vdecls) dvars
+  in (n'', vlist', Quantification q vdecls' form'' range)
+ Junction j sens r -> let
+   (n', vlist', sens') = foldl (\(x, vs, osens) s -> 
+                           let
+                            (x', vs', s') = mkDistVarsAux x vs s
+                           in (x', vs', s':osens)) (n, vlist, []) sens
+  in (n', vlist', Junction j (reverse sens') r)
+ Relation sen1 rel sen2 r -> let
+   (n1, vlist1, sen1') = mkDistVarsAux n  vlist  sen1
+   (n2, vlist2, sen2') = mkDistVarsAux n1 vlist1 sen2
+  in (n2, vlist2, Relation sen1' rel sen2' r)
+ Negation sen' r -> let 
+   (n', vlist', sen'') = mkDistVarsAux n vlist sen' 
+  in (n', vlist', Negation sen'' r)
+ _ -> (n, vlist, sen)
+       
+computePNF :: CASLFORMULA -> CASLFORMULA
+computePNF sen = case sen of 
+ Negation nsen _ -> let 
+   nsen' = computePNF nsen
+   in case nsen' of 
+       Quantification q vdecls qsen _ -> 
+         Quantification (dualQuant q) vdecls 
+                        (computePNF $ Negation qsen nullRange) nullRange
+       _ -> Negation nsen' nullRange
+ Relation sen1 Implication sen2 _ -> let 
+   sen1' = computePNF sen1
+   sen2' = computePNF sen2
+   in case sen1' of 
+      Quantification q vdecls qsen _ -> 
+        Quantification (dualQuant q) vdecls (computePNF $ Relation qsen Implication sen2 nullRange) nullRange
+      _ -> case sen2' of 
+        Quantification q vdecls qsen _ -> 
+          Quantification q vdecls 
+            (computePNF $ Relation sen1' Implication qsen nullRange) nullRange 
+          -- recursive call to catch multiple quantifiers for sen2'
+        _ -> -- no quantifications
+          Relation sen1' Implication sen2' nullRange
+ Relation sen1 RevImpl sen2 _ -> 
+  computePNF $ Relation sen2 Implication sen1 nullRange
+ Quantification q vdecls qsen _ -> 
+  Quantification q vdecls (computePNF qsen) nullRange
+ Junction j sens _ -> let
+   sens' = map computePNF sens
+   processQuants asen = case asen of
+     Quantification q vdecls sen' _ -> Quantification q vdecls (processQuants sen') nullRange
```

Comment:
Right, I wrote it thinking I'll call the function again for the quantified sentence, then it was not needed to do so and I didn't remove it. Will do so.